go/vt/vtctl/reparentutil/emergency_reparenter.go

```diff
@@ -0,0 +1,314 @@
+// perform EmergencyReparentShard operations using the given topo.Server,
+// TabletManagerClient, and logger.
+//
+// Providing a nil logger instance is allowed.
+func NewEmergencyReparenter(ts *topo.Server, tmc tmclient.TabletManagerClient, logger logutil.Logger) *EmergencyReparenter {
+	erp := EmergencyReparenter{
+		ts:     ts,
+		tmc:    tmc,
+		logger: logger,
+	}
+
+	if erp.logger == nil {
+		// Create a no-op logger so we can call functions on er.logger without
+		// needed to constantly check for non-nil.
+		erp.logger = logutil.NewCallbackLogger(func(*logutilpb.Event) {})
+	}
+
+	return &erp
+}
+
+// ReparentShard performs the EmergencyReparentShard operation on the given
+// keyspace and shard.
+func (erp *EmergencyReparenter) ReparentShard(ctx context.Context, keyspace string, shard string, opts EmergencyReparentOptions) (*events.Reparent, error) {
+	opts.lockAction = erp.getLockAction(opts.NewPrimaryAlias)
+
+	ctx, unlock, err := erp.ts.LockShard(ctx, keyspace, shard, opts.lockAction)
+	if err != nil {
+		return nil, err
+	}
+
+	defer unlock(&err)
+
+	ev := &events.Reparent{}
+	defer func() {
+		switch err {
+		case nil:
+			event.DispatchUpdate(ev, "finished EmergencyReparentShard")
+		default:
+			event.DispatchUpdate(ev, "failed EmergencyReparentShard: "+err.Error())
+		}
+	}()
+
+	err = erp.reparentShardLocked(ctx, ev, keyspace, shard, opts)
+
+	return ev, err
+}
+
+func (erp *EmergencyReparenter) getLockAction(newPrimaryAlias *topodatapb.TabletAlias) string {
+	action := "EmergencyReparentShard"
+
+	if newPrimaryAlias != nil {
+		action += fmt.Sprintf("(%v)", topoproto.TabletAliasString(newPrimaryAlias))
+	}
+
+	return action
+}
+
+func (erp *EmergencyReparenter) promoteNewPrimary(
+	ctx context.Context,
+	ev *events.Reparent,
+	keyspace string,
+	shard string,
+	newPrimaryTabletAlias string,
+	tabletMap map[string]*topo.TabletInfo,
+	statusMap map[string]*replicationdatapb.StopReplicationStatus,
+	opts EmergencyReparentOptions,
+) error {
+	erp.logger.Infof("promoting tablet %v to master", newPrimaryTabletAlias)
+	event.DispatchUpdate(ev, "promoting replica")
+
+	newPrimaryTabletInfo, ok := tabletMap[newPrimaryTabletAlias]
+	if !ok {
+		return vterrors.Errorf(vtrpc.Code_INTERNAL, "attempted to promote master-elect %v that was not in the tablet map; this an impossible situation", newPrimaryTabletAlias)
+	}
+
+	rp, err := erp.tmc.PromoteReplica(ctx, newPrimaryTabletInfo.Tablet)
+	if err != nil {
+		return vterrors.Wrapf(err, "master-elect tablet %v failed to be upgraded to master: %v", newPrimaryTabletAlias, err)
+	}
+
+	if err := topo.CheckShardLocked(ctx, keyspace, shard); err != nil {
+		return vterrors.Wrapf(err, "lost topology lock, aborting: %v", err)
+	}
+
+	replCtx, replCancel := context.WithTimeout(ctx, opts.WaitReplicasTimeout)
+	defer replCancel()
+
+	event.DispatchUpdate(ev, "reparenting all tablets")
+
+	// (@ajm188) - A question while migrating: Is this by design? By my read,
```

Comment:
@deepthi question for you in particular here. @setassociative this is the issue I was talking to you about the other day.

Reply:
It seems like an oversight. Either we care about errors from `handleReplica` and we should process them or not send them at all.